em++

```diff
@@ -11,6 +11,10 @@
 # To make modifications to this file, edit `tools/run_python_compiler.sh` and
 # then run `tools/create_entry_points.py`
 
+unset _PYTHON_SYSCONFIGDATA_NAME
+unset PYTHONHOME
+unset PYTHONPATH
```

Comment:
Do you need these two lines if you are using `-E` below?

Reply:
i think -E does not clear the env it just disregard the values, so better safe than sorry later be these 2 are really critical if another python child process is launched without -E

Reply:
OK, then in that case can you remove the `-E`?  There is no point in adding it if we have another method of doing that same thing and we think that we cannot inject `-E` into all the right place.  That would be `-E` as partial, and redundant solution here, right?

Maybe add a comment to that effect.  Something like "Unset these environment variables explicitly, rather than using `python -E`"

Reply:
-E covers (a lot!) of other variables that should not be impacted for emsdk, the two unset are really usefull failsafe for subprocesses where we I can't be sure -E will be set ( i did not read all emsdk source code).

tldr :
1) -E ensure fully proper operation of emsdk base tool called in first place.

2) the unset HOME/PATH ensure minimal proper operation of anything that could spawn as subprocesses of that emcc tool over those the shell script has no direct control.

Reply:
Do you know what else `-E` covers?   I guess the doc is quite open ended:

```
       -E     Ignore environment variables like PYTHONPATH and PYTHONHOME that modify the  behavior
              of the interpreter.
```

There are indeed a couple of places where emscripten spawns internal stuff using `sys.interpreter`.  With this change, those would run in partially-clean environment rather then that fully cleaned `-E` one?    It seems odd that `-E` would not also cover python sub-processes.   It also seems odd that we would consider partially-cleaned OK in some circumstances but not OK in others.

One example of emcc launching a python subprocess: https://github.com/emscripten-core/emscripten/blob/4549d9da3e85c96a71a7cf56de5e203b22801bfb/tools/building.py#L1435-L1440

Reply:
>  It seems odd that -E would not also cover python sub-processes

i'm quite sure -E does not impact subprocesses and there's a lot of ways to spawn them from python or other children. if you can be 100% certain to not miss one them, then ensure running sys.executable with -E would be of course better option at emsdk level.

But there's still a high probability that a non python child program starts another python regardless of -E, relying only on unsetting PYTHON* : i'm thinking of emconfigure, emmake and emcmake

Reply:
Does your use case really require `-E` or is removing those 3 environment variables sufficient to make it work?

Reply:
my case require both because we want to achieve good reproducibility if a user decide to build his own ci by downloading the sdk instead or using another CI than github, maybe want to upgrade emsdk without redownloading/rebuilding everything ( we don't want to patch like we do actualliy) , or use his custom python + custom env to drive emsdk that use his designed EMSDK_PYTHON.

Reply:
So what happens if run a tool like `emconfigure`, `emmake`, `emcmake`?    Lets say I have a tool that is run at configure time, and its written in python.

If I run `PYTHONPATH=/required/by/mytool ./configure` I would expect the python tool within my configure script to honor the `PYTHONPATH` I think?   So if `emconfigure` sanitizing those environment variables won't that break my expectations?

It seems that non-emscripten python scripts probably do want to honor those paths, even though emscirpten itself wants to be hermetic?

Reply:
Given that, I think the pure `-E` approach is probably best and we should work to ensure that emscripten tools are always launched via the launcher scripts (i've been thinking of adding a check for this actually to warn folks from inadvertently running `python emcc.py` themselves).

Reply:
Usually all system tools on good distribution should already set -E but again ... can't be 100% sure  ( it was worst in the past there are probably still remnants or obsolete systems that cannot upgrade for some reason )

PYTHONPATH or PYTHONHOME should never influence anything under configure/make/cmake ever and they should use a stable system python.

When the (auto)tools need to configure for **another python** they (should) provide a parameter to designate the targeted interpreter and run it for getting config data. This is where a given shell script wrapper should clear env then set PYTHON* and run the target correctly.

Reply:
But `PYTHONPATH=/required/by/mytool ./configure` does work today for cases where configure run some python script as part of configuring. 

This change would break that no?  Same for `PYTHONPATH=/required/by/mytool emmake`.

It doesnt seem like it should be our job to sanitize the environment for all our subprocesses, especially in the case of emmake/emcmake/emconfigure where the subprocess are use controlled and the user has a reasonable expectation that the environment they set will be passed to the children of make/cmake/configure.

So... I suggest we start with just using `-E`.. since that is clearly correct, and revisit that if it turns out that doesn't solve your issue?

Reply:
And maybe we should add `-E` to the windows `.bat` files for consistency?

Reply:
> But PYTHONPATH=/required/by/mytool ./configure does work today for cases where configure run some python script as part of configuring.

imho using PYTHON* that way is bad because configure is not a python script.

PYTHON* is for the end of the chain : the tool py script itself. not for influencing autotools that is thousands of lines that run thousands of tools .

> revisit that if it turns out that doesn't solve your issue

well actually it's not specifically my issue, it's more the various hard to solve potential issues that will come of thousands of modules around.

That we may hope to cross compile to wasm without patching emsdk ( pyodide still do that and iirc even rebuild it fully) and without having thousands of PR to sanitize subprocesses or helper tools we have not yet heard of.

 That in the case python-wasm-sdk come to interest more than just pygame ( less than  3minutes to build the wasm python module without docker, i'm sure it will get traction ! )


>  And maybe we should add -E to the windows .bat files for consistency?

i don't know anymore how system python runs on windows or mac systems (since 2.7), i think you should but i'm certainly not the one to ask if you can safely.

Reply:
> imho using PYTHON* that way is bad because configure is not a python script.

It might be in-advised but that doesn't mean we should break anyone who want to use it.

> well actually it's not specifically my issue, it's more the various hard to solve potential issues that will come of thousands of modules around.
> 
> That we may hope to cross compile to wasm without patching emsdk ( pyodide still do that and iirc even rebuild it fully) and without having thousands of PR to sanitize subprocesses or helper tools we have not yet heard of.

I think that number of python subprocess that get run by emcc should be very limited, we should be able to keep track of them and ensure they all get run with -E.    

> i don't know anymore how system python runs on windows or mac systems (since 2.7), i think you should but i'm certainly not the one to ask if you can safely.

I think we should add `-E` everywhere (including on windows) for consistency.   It could be confusing otherwise.  We have a fair amount of CI testing to ensure it works, so you don't need to worry about locally testing it.